### PR TITLE
Rework retransmission buffer

### PR DESCRIPTION
**Pull request by Hartigan**
Main goals:
1. Fix memory leak of cloned RTP packets;
2. Reduce allocations in retransmission buffer

### CODE REVIEW

**Comment by nazar-pc** [comment]
Thank you! I'll check when I get time, should be this week.

**Comment by jmillan** [comment]
@Hartigan, do you know where exactly was the leak?

**Comment by Hartigan** [comment]
@jmillan I found leak of cloned rtp packets(created for '''StorageItemBuffer'''). Then I tried to figure out how buffer works, but it's hard for me. And I just rewrote it in simpler way with fixed addressing('''seq = (bucketIdx, inBucketPos)'''). I think something wrong with translation from '''seq''' to deque '''idx'''

**Comment by Hartigan** [comment]
@jmillan link to massif dump with leak: https://gist.github.com/Hartigan/b77568249e1f5b0f95c0e5db7df5bc97

File: worker/include/RTC/RtpStreamSend.hpp [hunk 1]
```diff
@@ -36,23 +36,43 @@ namespace RTC
 			bool rtxEncoded{ false };
 		};
 
+		class Bucket {
```
Comment (jmillan) [inline worker/include/RTC/RtpStreamSend.hpp]: This seems to be a good candidate to be a template class accessible as an utility. In this concreate case it would be a `Bucket` of `StorageItem`.

Comment (Hartigan) [reply worker/include/RTC/RtpStreamSend.hpp]: @jmillan  smth like ```StorageItemBuffer<TKey, TValue, BucketSize>``` and ```StorageItemBuffer<uint16_t, StorageItem, 256>```?

Comment (Hartigan) [reply worker/include/RTC/RtpStreamSend.hpp]: I'll prepare commit only for ```Bucket<T>```

File: worker/src/RTC/RtpStreamSend.cpp [hunk 2]
```diff
@@ -36,114 +37,198 @@ namespace RTC
...
+		auto item = this->buffer[pos];
+		this->buffer[pos] = nullptr;
+
+		resetStorageItem(item);
+		StorageItemPool.Return(item);
+	}
+
+	void RtpStreamSend::Bucket::Clear()
+	{
+		for(size_t pos = 0; pos < RtpStreamSend::Bucket::BucketSize && !this->IsEmpty(); ++pos)
```
Comment (jmillan) [inline worker/src/RTC/RtpStreamSend.cpp]:
```suggestion
		for(size_t pos = 0; pos < this->buffer.size() && !this->IsEmpty(); ++pos)
```

**Comment by jmillan** [comment]
just wondering,

~~Why having buckets implemented this way instead of using `std::unordered_map` which already implement those? In unordered maps search, insertion, and removal of elements have average constant-time complexity.~~

Please disregard.

**Comment by Hartigan** [comment]
@jmillan Main reason is optimize search of oldest ```StorageItem``` after remove current oldest.
In ```std::unordered_map``` we need ```O(n)``` operations in bad case. With this implementation ```O(2*sqrt(n))``` in bad case.
Bad case: packets received out-of-order or sequence overflow max ```uint16_t```

File: worker/include/RTC/RtpStreamSend.hpp (hunk 1 above)
Comment (Hartigan) [reply worker/include/RTC/RtpStreamSend.hpp]: Moved into ```Utils.hpp```

File: worker/src/RTC/RtpStreamSend.cpp (hunk 2 above)
Comment (Hartigan) [reply worker/src/RTC/RtpStreamSend.cpp]: Updated. ```Bucket::Clear()``` moved to ```StorageItemBuffer::Clear()``` because ```StorageItemPool``` located in ```RtpStreamSend```

File: worker/src/RTC/RtpStreamSend.cpp [hunk 3]
```diff
@@ -36,107 +37,162 @@ namespace RTC
...
+			return this->Get(static_cast<uint16_t>(this->oldestSeq));
 		}
-		// If we have erased the last element, remove all `nullptr` elements from the end of the buffer.
-		else if (idx == static_cast<uint16_t>(this->buffer.size() - 1))
+
+		return nullptr;
+	}
+
+	void RtpStreamSend::StorageItemBuffer::TryUpdateOldest(uint16_t seq, RtpStreamSend::StorageItem* storageItem)
+	{
```
Comment (jmillan) [inline worker/src/RTC/RtpStreamSend.cpp]: Considering this code is only called by  `Insert()` and it is so tight to it and small, does it make sense having a separate method for this?

File: worker/src/RTC/RtpStreamSend.cpp [hunk 4]
```diff
@@ -36,107 +37,162 @@ namespace RTC
...
+		if (!oldest)
 		{
-			while (!this->buffer.back())
-				this->buffer.pop_back();
+			this->oldestSeq = seq;
+			return;
 		}
 
-		return true;
+		if (oldest->originalPacket->GetTimestamp() > storageItem->originalPacket->GetTimestamp())
```
Comment (jmillan) [inline worker/src/RTC/RtpStreamSend.cpp]: We cannot compare timestamps this way since the value can be wrapped around. Use this:

```c++
RTC::SeqManager<uint32_t>::IsSeqHigherThan(oldest->originalPacket->GetTimestamp(), storageItem->originalPacket->GetTimestamp())
```

Comment (Hartigan) [reply worker/src/RTC/RtpStreamSend.cpp]: done

File: worker/src/RTC/RtpStreamSend.cpp (hunk 3 above)
Comment (Hartigan) [reply worker/src/RTC/RtpStreamSend.cpp]: I think it's part of my code-style :D I inlined this.

**Comment by jmillan** [comment]
Please run `npm run format:worker` to fix formatting.

Edit: you need to run `npm run install-clang-tools` first.

File: worker/include/RTC/RtpStreamSend.hpp [hunk 5]
```diff
@@ -45,14 +48,19 @@ namespace RTC
...
 
 		private:
-			uint16_t startSeq{ 0 };
-			std::deque<StorageItem*> buffer;
+			void InvalidateOldest(uint16_t prevOldest);
+			static size_t GetBucketIndex(uint16_t seq);
+			static size_t GetPositionInBucket(uint16_t seq);
+
+			int32_t oldestSeq{ -1 };
+			std::array<Utils::Bucket<StorageItem, BucketSize>*, BucketsCount> buckets{};
```
Comment (nazar-pc) [inline worker/include/RTC/RtpStreamSend.hpp:63]: I don't understand the idea behind 2 arrays/levels of indirection, can you explain why is this needed?

Comment (Hartigan) [reply worker/include/RTC/RtpStreamSend.hpp:63]: @nazar-pc It's working like [virtual memory](https://en.wikipedia.org/wiki/Virtual_memory). I created [visualization](https://gist.github.com/Hartigan/61e194fcd561c14c0e2b8086327034f9). ```Buckets/Pages``` are reusable in thread.

Comment (nazar-pc) [reply worker/include/RTC/RtpStreamSend.hpp:63]: I think I understand what, I don't understand why though. It worked fine before (minus apparent bug somewhere) and used much less memory than this approach after quick glance.

Comment (Hartigan) [reply worker/include/RTC/RtpStreamSend.hpp:63]: ```std::deque``` allocations depends on [platform](https://bugs.chromium.org/p/chromium/issues/detail?id=674287). For ```MSVC``` it's working like linked list with frequent allocation. My approach uses buffers with size ```8 * 256 = 2048 bytes``` and ```8 * 256 = 2048 bytes``` for bucket's ```std::array```.
```std::deque``` buffer sizes :
- MSVC: 16 bytes or element size if it's bigger(16 bytes for our case)
- GCC: 512 bytes or element size if it's bigger(512 bytes for our case)
- Clang: element_size < 256 ? 4096 : element_size * 16(4096 bytes for our case).

So, my approach uses comparable memory size with ```std::deque``` for GCC/CLang, but consumes much less allocations.

Also, ```std::deque``` it's linked list of buffers, so you need to dereference the pointer multiple times for ```Get``` by index(depends on buffers number and implementation). In my approach we need only 2 dereference operations.

Comment (nazar-pc) [reply worker/include/RTC/RtpStreamSend.hpp:63]: Oh wow, this is such an inefficient implementation in C++... In Rust it is basically the same as vector just with extra pointer. I'd rather use vector then and resize it as necessary. Having virtual memory implementation feels like overkill, all we need is a flat data structure with ability to point to a fixed location in it and rarely grow. If we do this we'll save both some memory and and access time on indirections.

Comment (nazar-pc) [reply worker/include/RTC/RtpStreamSend.hpp:63]: https://github.com/Shmoopty/veque looks promising, I can write `meson.build` for it.

Comment (Hartigan) [reply worker/include/RTC/RtpStreamSend.hpp:63]: ```Boost``` provides [circular_buffer_space_optimized](https://www.boost.org/doc/libs/1_78_0/doc/html/circular_buffer.html), but need to implement ```seq``` to ring buffer ```idx``` translation and import ```boost```.

Comment (Hartigan) [reply worker/include/RTC/RtpStreamSend.hpp:63]: Anyway, current implementation on ```deque``` has not obvious translation from ```seq``` to in ```deque``` position and may contain hidden bugs.

**Comment by Hartigan** [comment]
@nazar-pc  At nigth I created new massfi dump. And found packet leak in that MR. After investigatetions, I found bug  in both implementations.
```
// Cleanup is finished if we found an item with recent enough packet, but also account
// for out-of-order packets.
if (diffMs < this->retransmissionBufferSize || packetTs < checkedPacketTs)
    break;
```

Wrong comparsion of timestamp prevents cleaning of buffer.

**Comment by nazar-pc** [comment]
Thank you! I'll take a look when I have time.

**Comment by jmillan** [comment]
Out of curiosity,

> @jmillan Main reason is optimize search of oldest StorageItem after remove current oldest.
In std::unordered_map we need O(n) operations in bad case. With this implementation O(2*sqrt(n)) in bad case.
Bad case: packets received out-of-order or sequence overflow max uint16_t

Does this also apply considering we can provide a Key Hash function to `unordered_map`?

```c++
#include <iostream>
#include <unordered_map>
#include <cstdint>

const size_t bucketCount = 256;

struct KeyHash {
 std::size_t operator()(const uint16_t& key) const
 {
     return key / bucketCount;
 }
};

int main()
{
     std::unordered_map<uint16_t, std::string, KeyHash> map = {
            { 1, "foo"},
            { 260, "bar"}
    };
    
    std::cout<< "bucket(1): " << map.bucket(1) << std::endl; // 0
    std::cout<< "bucket(266): " << map.bucket(260) << std::endl; // 1
    
    return 0;
}
```

**EDIT:** Getting the specific item from the bucket wouldn't still be as efficient.

**Comment by Hartigan** [comment]
@jmillan  @nazar-pc 
I tests again:
This PR:
1. [2 hours test](https://gist.github.com/Hartigan/18d23fc858d9c9eea85cf360c902cbdc)
2. [2 hours test](https://gist.github.com/Hartigan/b7173f6cc7f6d3fc0285bb30589426d1)
3. [5 hours test](https://gist.github.com/Hartigan/72ef3a34c37c1100e1841f66d35a4a40)
And I haven't reproduced the memory leak yet. I gonna try longer tests.

Also I tried fix ```timestamp``` comparisons and ```seq``` calculations in current implementations: [10 hours test](https://gist.github.com/Hartigan/c9f1bbc24cd44daa32af3b0ea4670573) . But this is a failed attempt.

**Comment by nazar-pc** [comment]
Do you mean latest tip of the memory-optimizations branch (commit af3b681f1355835931c9d4a3d960cbe0a42e739f or newer) is still leaking?

**Comment by Hartigan** [comment]
@nazar-pc  It's outdated i think
